another solution to apt_repository/repos_urls traceback on Debian (PR #1738, bug #1684)

amenonsen opened the pull request:
This fixes the `i[3]` reference in one commit, and adds a Debian version of `repos_urls` without ppa handling. This does work on Debian and should work on Ubuntu (since the code hasn't changed there, except to call `_expand_ppa` with the right arguments).

pilou- committed:
fix error occurring with Debian

Error was: AttributeError: 'SourcesList' object has no attribute 'repos_urls'

amenonsen committed:
Fix call to _expand_ppa

amenonsen committed:
Clarify HAVE_PYTHON_APT/install_python_apt handling in apt_repository

1. Don't test check_mode in both the caller and in the callee.
2. Don't test HAVE_PYTHON_APT inside an if that tests HAVE_PYTHON_APT
3. Don't be irritatingly vague about why the module fails ("You may be
   seeing this because…").

Note that if «apt-get -y install python-apt» succeeds with rc==0, but
for some reason python_apt is not usable afterwards, this will break
because the imports in install_python_apt aren't wrapped inside a
try/except.

In other words, we assume that install_python_apt either succeeds or
fails with a traceback. This commit doesn't affect that behaviour.

amenonsen commented:
I pushed an additional commit to this PR to clean up the unnecessarily convoluted HAVE_PYTHON_APT logic (there were double-tests in multiple places, and an annoyingly vague error message). It's tested on Debian and works fine. This can be merged on its own too, if desired.

amenonsen commented:
And on further consideration, here's another commit that adds a Debian-specific version of `_expand_ppa` that always fails. Because otherwise, if you actually tried to add a ppa repository on Debian, you would get the following traceback because of the call in `main`:

```
AttributeError: 'SourcesList' object has no attribute '_expand_ppa'
```

After this change, here's what I get:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "ppa repositories are supported only on Ubuntu
```

This implements the behaviour that @sashka seemed to intend for #1738, i.e. that adding ppas fails cleanly on Debian. @abadger does this look OK to you? It works on Debian for me.

/cc @pilou- @verm666 @dalyc does this work for you?

amenonsen commented:
See https://github.com/amenonsen/ansible-modules-core/tree/repounsplit for an alternative implementation that removes the `UbuntuSourcesList` class altogether. This works with both normal repo lines and ppas on Debian for me. I think this implementation is actually cleaner and more understandable than what we had earlier.

@abadger, I'll await your verdict, if you prefer to commit the single-class version, let me know and I can submit that as a separate PR and close this one. Or if you want to stick to the intention of the original code as @sashka described it, then this PR is the conservative option that still does a bit of worthwhile cleanup.

pilou- commented:
A PPA is just an APT repository: adding a PPA repository on Debian should work.

sashka commented:
PPA is an APT repository, `ppa:` shortcut is not.

pilou- commented:
@sashka indeed, #1835 was not that bad ;)

amenonsen committed:
Make SourcesList __init__ method also set self.module

This was originally required to allow other methods in SourcesList to
fail, but subsequent changes rendered that unnecessary, and it's just
a cleanup now, and avoids passing in module separately to save().

amenonsen committed:
Simplify distribution test

If it's Ubuntu, use UbuntuSourcesList; if it's any other apt-friendly
distribution, use SourcesList; otherwise, fail.

amenonsen commented:
@pilou- in fact, I think #1835 was better than what I did. Since your PR was already closed, I included your commit in this PR instead of my earlier solution. @sashka note that @pilou-'s solution doesn't need to put any `repos_urls` or `expand_ppa` into the base class at all. I look forward to any comments. For the record, trying a `ppa:` repo string on Debian now gives this error:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Invalid repository string: ppa:nginx/stable"}
```

sashka commented:
:+1:

I also passed it through simple testing like this: https://gist.github.com/sashka/f4f6704224f4fc1e8a98
It works like a charm.

abadger commented:
Thanks @amenonsen,  @verm666, and @pilou- for implementations and @sashka for reviewing!  glad to get this merged.

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
